test/jdk/java/nio/channels/Selector/DeferredClose/java.base/java/net/InetSocketAddress.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package java.net;
+
+
+import java.util.Locale;
+
+// Patched implementation only meant to be used in certain tests
+public class InetSocketAddress extends SocketAddress {
+
+    @java.io.Serial
+    private static final long serialVersionUID = 5076001401234631237L;
+
+    public static boolean enableDelay;
```

Comment:
should it be volatile?

Reply:
The main thing for this test is that it's easy to maintain because someone looking at the test in a few years (maybe because of a test failure) will want to understand it quickly. I assume it doesn't need to be a public static field, it could be a private instance field. I think that would remove the need for the beforeAll/afterAll as it would allow the test to just set the delay for the ISA used by the test. I don't want to water Jai's time on too many iterations on this test, I think the main thing is that it is reliable and easy to maintain.

Reply:
Hello Daniel, I don't think it's needed. We set this to `true` (only) once in a `@BeforeAll` which gets called by one single thread. The subsequent multiple threads will only get created and launched as part of the test methods after the `@BeforeAll` has finished, so I believe they would all correctly see enableDelay as true.

After that, we only update that `enableDelay` value after all tests are done, so that too should be fine.

Reply:
> I assume it doesn't need to be a public static field, it could be a private instance field.

It should have been a private field. It's an oversight that I left it as public in a recent update. I'll fix that.

> I think that would remove the need for the beforeAll/afterAll as it would allow the test to just set the delay for the ISA used by the test.

The important detail here is that to consistently reproduce this issue both in send and receive methods of the DatagramChannelImpl, the delay has to be introduced not just in the ISA constructed by the test methods, but also in the ISA constructed by the DatagramChannelImpl implementation's decoding of native addresses to ISA. That decoding of ISA happens deep within the receive operation after it has acquired the relevant lock and that's the place where we want to slow things down so that a concurrent thread can then enter and finish close the channel which was in deferred close state.

If we wanted to reproduce the issue only for send, then constructing the delayed ISA in the test method should have been enough (because send would just call the ISA.getPort() on the ISA passed by the test method). But I also wanted to be sure that there was a way to reproduce (and thus verify the fix) for receive too (which internally calls the ISA constructor at least once upon receiving a datagram). That was the reason why I chose to have this in the beforeAll/afterAll, to keep it relatively simple.

Reply:
Okay, I think we can live with that.

Reply:
Technically - it may be set in only one thread but it's consulted in many other threads. However there should be already enough full memory barreers around to ensure that the other threads will see the modifications, especially since these threads are started after the modification was made. So it shouldn't be an issue for the test in this form.